### PR TITLE
Initial Fix

### PR DESCRIPTION
**Pull request by loongyh**
Haven't initialized the adafruit FP lib yet.

### CODE REVIEW

File: rxxx.cpp [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+void RxxxComponent::finish_enrollment(uint8_t result) {
+  this->enrollment_callback_.call(result == FINGERPRINT_OK, result, enrollmentSlot_);
+  enrollmentImage_ = 0;
+  enrollmentSlot_ = 0;
+  enrolling_binary_sensor_->publish_state(false);
+}
+
+void RxxxComponent::scan_and_match() {
+  uint8_t result = scan_image(1);
+  int finger_id = -1;
```
Comment (OnFreund) [inline rxxx.cpp]: Maybe we can initialize to zero to make this unsigned? AFAIK zero is not a valid finger id so it should be ok.

File: rxxx.cpp [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+  uint16_t confidence = 0;
+  if (result == FINGERPRINT_NOFINGER) {
+    return;
+  }
+  if (result == FINGERPRINT_OK) {
```
Comment (OnFreund) [inline rxxx.cpp:170]: Might be better to put `waitingRemoval = true;` before this line. That way if there's some exception afterwards, we're not stuck in an error loop as long as the finger is on the reader.

**Review by OnFreund — COMMENTED** [review]
Looks good.
As for using my changed library in PlatformIO, I know that it's doable (`OnFreund/Adafruit-Fingerprint-Sensor-Library` in the YAML works), but not sure how it's done from a custom component.

**Comment by loongyh** [comment]
> `OnFreund/Adafruit-Fingerprint-Sensor-Library` in the YAML works

I believe it's best if we can import the lib directly in the codegen so that's one less extra step and yaml line for the user.

Have you seen [my reply](https://github.com/loongyh/esphome-fingerprint/issues/7#issuecomment-708965358) regarding submitting a PR to adafruit?

File: rxxx.cpp (hunk 1 above)
Comment (loongyh) [reply rxxx.cpp]: Yep my thoughts exactly :) will do

**Comment by OnFreund** [comment]
> I believe it's best if we can import the lib directly in the codegen

I agree, but I'm assuming the same syntax should work there too.

> Have you seen my reply regarding submitting a PR to adafruit?

Created the PR here: https://github.com/adafruit/Adafruit-Fingerprint-Sensor-Library/pull/89

**Comment by loongyh** [comment]
Adafruit just merged your stream support, till they release a new version we'd have to use the git url for now. Have updated the PR to initialize the lib.
```
# rxxx.yaml

uart:
  tx_pin: 17
  rx_pin: 16
  baud_rate: 57600

rxxx:
  sensing_pin: 4

binary_sensor:
  - platform: rxxx
    enrolling:
      name: "Enrolling"

sensor:
  - platform: rxxx
    fingerprint_count:
      name: "Fingerprint Count"
    status:
      name: "Status"
    capacity:
      name: "Capacity"
    security_level:
      name: "Security Level"
    last_finger_id:
      name: "Last Finger ID"
    last_confidence:
      name: "Last Confidence"
```




Also added in the following icons in const.py:
```
# esphome/const.py

ICON_ACCOUNT = 'mdi:account'
ICON_ACCOUNT_CHECK = 'mdi:account-check'
ICON_DATABASE = 'mdi:database'
ICON_FINGERPRINT = 'mdi:fingerprint'
ICON_SECURITY = 'mdi:security'
```

**Comment by OnFreund** [comment]
Cool!

How is Last Finger ID zero, though? It's supposed to be set to a positive number of matching a finger.
After we fix this, maybe we can open a PR against ESPHome, and continue there?

**Comment by loongyh** [comment]
I used the [SFGDemo](http://www.adafruit.com/datasheets/SFGDemoV2.0.rar) program to enrol and it seems they start from 0, whereas the adafruit lib starts from 1. Will investigate further.

**Comment by OnFreund** [comment]
The adadfuit lib doesn't really "start" from anywhere, they just pass the slot on to the reader without modification. If slot 0 can be used, we might need to initialize `finger_id` to something else in `scan_and_match()`.

**Comment by OnFreund** [comment]
Don't know if you saw this, but AdaFruit released version 2.0.4 with the stream constructor.

**Comment by loongyh** [comment]
> The adadfuit lib doesn't really "start" from anywhere, they just pass the slot on to the reader without modification.

Ah yes, I meant in the example enroll.ino in which for some reason `ID #0 not allowed`.

> AdaFruit released version 2.0.4

Nice! Will update the codegen to use it.

**Comment by OnFreund** [comment]
If 0 is an allowed slot, let's switch to a signed int16 so we can use -1 for an invalid print. It'll still have over 32k positive indexes. Even if you find a reader that supports so many, and enroll a print in just 10 seconds, it will take almost 4 straight days with no breaks just to enroll that many :)

**Comment by loongyh** [comment]
Hmm I kinda think there's a legit reason the example excluded slot #0.

**Comment by OnFreund** [comment]
> Hmm I kinda think there's a legit reason the example excluded slot #0.

True, but then again you've showed that you can successfully use index 0...

File: rxxx.cpp [hunk 3]
```diff
@@ -34,13 +34,15 @@ void RxxxComponent::update() {
...
 
-  if (this->sensing_pin_->digital_read() == HIGH) {
-    ESP_LOGV(TAG, "No touch sensing");
-    return;
+  if (this->sensing_pin_ != nullptr) {
+    if (this->sensing_pin_->digital_read() == HIGH) {
+      ESP_LOGV(TAG, "No touch sensing");
+      return;
+    }
   }
```
Comment (OnFreund) [inline rxxx.cpp]: Maybe use `&&` instead of nested `if`s?

File: rxxx.cpp [hunk 4]
```diff
@@ -34,13 +34,15 @@ void RxxxComponent::update() {
...
 
   if (this->enrollment_image_ == 0) {
-    ESP_LOGD(TAG, "Scan and match");
+    ESP_LOGV(TAG, "Scan and match");
```
Comment (OnFreund) [inline rxxx.cpp]: When sensing pin is available, this happens infrequently enough to make it `debug` rather than `verbose`

File: rxxx.cpp [hunk 5]
```diff
@@ -109,10 +111,10 @@ void RxxxComponent::scan_and_match_() {
 }
 
 uint8_t RxxxComponent::scan_image_(uint8_t buffer) {
-  ESP_LOGD(TAG, "Getting image %d", buffer);
+  ESP_LOGV(TAG, "Getting image %d", buffer);
```
Comment (OnFreund) [inline rxxx.cpp]: Sames here

**Review by OnFreund — CHANGES_REQUESTED** [review]

**Comment by loongyh** [comment]
> True, but then again you've showed that you can successfully use index 0...

Have implemented back slot #0. Just finished up the initial fix, passed linting and tested working with my full config as follows:

```
# rxxx.yaml
api:
  reboot_timeout: 0s

uart:
  tx_pin: 17
  rx_pin: 16
  baud_rate: 57600

rxxx:
  sensing_pin: 4
  on_finger_scanned:
    - lambda: |-
        if (success) {
          id(rxxx_state).publish_state("Authorized finger " + to_string(finger_id));
        } else {
          if (result == FINGERPRINT_NOTFOUND) {
            id(rxxx_state).publish_state("Unauthorized finger");
          } else {
            id(rxxx_state).publish_state("Error: " + to_string(result));
          }
        }   
  on_enrollment_scan:
    - lambda: |-
        id(rxxx_state).publish_state("Scanned " + to_string(scan_number));
  on_enrollment_done:
    - lambda: |-
        if (success) {
          id(rxxx_state).publish_state("Enrolled into slot " + to_string(finger_id));
        } else {
          id(rxxx_state).publish_state("Error: " + to_string(result));
        } 

binary_sensor:
  - platform: rxxx
    id: rxxx_enrolling
    name: "Rxxx Enrolling"

text_sensor:
  - platform: template
    id: rxxx_state
    name: "Rxxx State"

sensor:
  - platform: rxxx
    fingerprint_count:
      name: "Rxxx Fingerprint Count"
    status:
      name: "Rxxx Status"
    capacity:
      name: "Rxxx Capacity"
    security_level:
      name: "Rxxx Security Level"
    last_finger_id:
      name: "Rxxx Last Finger ID"
    last_confidence:
      name: "Rxxx Last Confidence"
  - platform: homeassistant
    id: hass_input_num
    entity_id: input_number.rxxx
    accuracy_decimals: 0

switch:
  - platform: template
    name: "Rxxx Enroll"
    icon: "mdi:account-plus"
    lambda: !lambda "return id(rxxx_enrolling).state;"
    turn_on_action:
      - rxxx.enroll:
          finger_id: !lambda "return id(hass_input_num).state;"
    turn_off_action:
      - rxxx.cancel_enroll:
  - platform: template
    name: "Rxxx Clear DB"
    icon: "mdi:nuke"
    turn_on_action:
      - rxxx.delete_all:

```

```
# HA configuration.yaml
input_number:
  rxxx:
    name: Slot No.
    icon: mdi:content-save
    min: 0
    max: 200
    mode: box
```

```
# HA ui-lovelace.yaml
views:
  - title: test
    icon: mdi:sign-caution
    cards:
      - type: horizontal-stack
        cards:
        - type: entities
          show_header_toggle: false
          entities:
            - switch.rxxx_enroll
            - input_number.rxxx
            - binary_sensor.rxxx_enrolling
            - sensor.rxxx_state
            - sensor.rxxx_last_finger_id
            - sensor.rxxx_last_confidence
            - sensor.rxxx_fingerprint_count
            - sensor.rxxx_capacity
            - switch.rxxx_clear_db
            - sensor.rxxx_status
            - sensor.rxxx_security_level
```




> After we fix this, maybe we can open a PR against ESPHome, and continue there?

Still need to write up the docs first before we can open the PR, also we have to do it not with this current component-only repo but with the full fork, taking note of the icon additions in `const.py`:
```
# const.py
ICON_ACCOUNT = 'mdi:account'
ICON_ACCOUNT_CHECK = 'mdi:account-check'
ICON_DATABASE = 'mdi:database'
ICON_FINGERPRINT = 'mdi:fingerprint'
ICON_KEY_PLUS = 'mdi:key-plus'
ICON_SECURITY = 'mdi:security'
```

File: rxxx.cpp (hunk 4 above)
Comment (loongyh) [reply rxxx.cpp]: Do you want to include a conditional check to only logd when sensing pin is available?

File: rxxx.cpp (hunk 3 above)
Comment (loongyh) [reply rxxx.cpp]: Okay

File: rxxx.cpp (hunk 4 above)
Comment (loongyh) [reply rxxx.cpp]: Added the conditional checks

**Comment by loongyh** [comment]
Have committed the last of my changes, closed all the issues and it's all ready now :) Can you review and proceed to submit the PR to ESPHome? I'll be unable to work on this for the next 2 weeks starting tomorrow.

**Comment by loongyh** [comment]
Latest configs as follows:

```
# rxxx.yaml
api:
  reboot_timeout: 0s

uart:
  tx_pin: 17
  rx_pin: 16
  baud_rate: 57600

rxxx:
  sensing_pin: 4
  on_finger_scan_matched:
    - rxxx.aura_led_control:
        state: BREATHING
        speed: 200
        color: BLUE
        count: 1
    - text_sensor.template.publish:
        id: rxxx_state
        state: "Authorized finger"
  on_finger_scan_unmatched:
    - rxxx.aura_led_control:
        state: FLASHING
        speed: 25
        color: RED
        count: 2
    - text_sensor.template.publish:
        id: rxxx_state
        state: "Unauthorized finger"
  on_enrollment_scan:
    - rxxx.aura_led_control:
        state: FLASHING
        speed: 25
        color: BLUE
        count: 2
    - rxxx.aura_led_control:
        state: ALWAYS_ON
        speed: 0
        color: PURPLE
        count: 0
  on_enrollment_done:
    - rxxx.aura_led_control:
        state: BREATHING
        speed: 100
        color: BLUE
        count: 2
    - text_sensor.template.publish:
        id: rxxx_state
        state: "Enrolled fingerprint"
  on_enrollment_failed:
    - rxxx.aura_led_control:
        state: FLASHING
        speed: 25
        color: RED
        count: 4
    - text_sensor.template.publish:
        id: rxxx_state
        state: "Failed to enroll fingerprint"

binary_sensor:
  - platform: rxxx
    id: rxxx_enrolling
    name: "Rxxx Enrolling"

text_sensor:
  - platform: template
    id: rxxx_state
    name: "Rxxx State"

sensor:
  - platform: rxxx
    fingerprint_count:
      name: "Rxxx Fingerprint Count"
    last_finger_id:
      name: "Rxxx Last Finger ID"
    last_confidence:
      name: "Rxxx Last Confidence"
  - platform: homeassistant
    id: hass_input_num
    entity_id: input_number.rxxx
    accuracy_decimals: 0

switch:
  - platform: template
    name: "Rxxx Enroll"
    icon: "mdi:account-plus"
    lambda: !lambda "return id(rxxx_enrolling).state;"
    turn_on_action:
      - rxxx.aura_led_control:
          state: ALWAYS_ON
          speed: 0
          color: PURPLE
          count: 0
      - rxxx.enroll: !lambda "return id(hass_input_num).state;"
    turn_off_action:
      - rxxx.cancel_enroll:
  - platform: template
    name: "Rxxx Delete Fingerprint"
    icon: "mdi:account-minus"
    turn_on_action:
      - rxxx.delete: !lambda "return id(hass_input_num).state;"
  - platform: template
    name: "Rxxx Clear DB"
    icon: "mdi:nuke"
    turn_on_action:
      - rxxx.delete_all:
```

```
# HA configuration.yaml
input_number:
  rxxx:
    name: Slot No.
    icon: mdi:content-save
    min: 0
    max: 200
    mode: box
```

```
# HA ui-lovelace.yaml
views:
  - title: test
    icon: mdi:sign-caution
    cards:
      - type: horizontal-stack
        cards:
        - type: entities
          show_header_toggle: false
          entities:
            - switch.rxxx_enroll
            - input_number.rxxx
            - binary_sensor.rxxx_enrolling
            - sensor.rxxx_state
            - sensor.rxxx_last_finger_id
            - sensor.rxxx_last_confidence
            - sensor.rxxx_fingerprint_count
            - switch.rxxx_delete_fingerprint
            - switch.rxxx_clear_db
```

**Comment by zurpia** [comment]
Great, thanks @loongyh and @OnFreund!. First time testing components on ESPHOME. Any advice? I have used it many times but never tested a new component. Hardware is ready.

**Comment by loongyh** [comment]
1. Follow the [Contributing guide](https://esphome.io/guides/contributing.html#setting-up-development-environment) to set up the esphome dev environment, then simply clone our repo under the esphome/components directory.
If you have an existing version of esphome installed already, then you can use python virtualenv to containerize it:
   ```
   python3 -m venv ~/esphome-dev
   source ~/esphome-dev/bin/activate
   pip3 install wheel
   git clone https://github.com/esphome/esphome.git
   cd ~/esphome-dev/esphome/script
   ./setup
   cd ~/esphome-dev/esphome/esphome/components
   git clone https://github.com/loongyh/esphome-fingerprint.git
   mv esphome-fingerprint rxxx
   ```

2. Refer above for the icons to add in ~/esphome-dev/esphome/esphome/const.py, as well as my yaml config examples

3. Whenever you want to compile against this dev build, `source ~/esphome-dev/bin/activate` first and verify the virtualenv is active with the (esphome-dev) shell prompt prefix. You can also check with `esphome version` command, it should show dev version.

**Comment by loongyh** [comment]
@OnFreund @zurpia Written up the [docs](https://github.com/loongyh/esphome-docs/tree/rxxx-docs), please do review!

**Comment by loongyh** [comment]
@OnFreund I did a few changes during my pr submission to esphome, this is outdated

**Comment by OnFreund** [comment]
No worries. Hopefully soon the ESPHome repo would be the canonical source for this anyway :)

**Comment by loongyh** [comment]
Yep, let's continue from there 😁

**Comment by hungdo0993** [comment]
Hello guy, i am testing this component, i have 1 question, the connection for the sensing_pin. Is it the wake up signal (blue wire) from the sensor?

**Comment by OnFreund** [comment]
@hungdo0993 That's correct

**Comment by hungdo0993** [comment]
Thank you for quick reply

**Comment by hungdo0993** [comment]
Thank for your component, i have tested it already. Everything is working fine.

**Comment by DirkTas67** [comment]
Hello,

Where can I find the information about how to connect this to an ESP8266/ESP32.

Regards,

**Comment by OnFreund** [comment]
That's not part of esphome yet. You can follow the progress here:
https://github.com/esphome/esphome/pull/1356

**Comment by harveyyoyo** [comment]
So I got this working with the grow fingerprint component. My issue is that it doesnt pick up the fingerprint most of the time. It just keeps saying no finger. It goes through occasionally. With the arduino adafruit it goes all the time. So there must be something causing this issue.

**Comment by loongyh** [comment]
@harveyyoyo
1. Which model are you using?
2. Are you using the sensing pin?
3. Did the 'no finger' appear once every second, or does it only appear once after you place your finger on the reader?

**Comment by harveyyoyo** [comment]
Im using the R307 which does not have a sensing pin. im using a nodemcu as well.
The no finger appears multiple times per second.
OK i just realized I had the sensing pin listed and now removed it so it doesnt keep saying no finger
its still not great on picking up my finger. many times i place my finger there it just doesnt register that my finger is there. it seems moody, which could also mean i have a bad sensor. idk. I never used actually used it with the adafruit code i just tested it that way.  If you have any ideas let me know.
I also have a question. last finger id will always have the last finger print listed. how do i create an unlock automation based on last fingerprint.

**Comment by loongyh** [comment]
> which could also mean i have a bad sensor.

Do you have another reader to try?

> I also have a question. last finger id will always have the last finger print listed. how do i create an unlock automation based on last fingerprint.

There's an example config to unlock on matching scan written under the documentation [here](https://next.esphome.io/components/fingerprint_grow.html). If you're comfortable with lambdas, you can use the variable `finger_id` in the `on_finger_scanned_matched` trigger to do something based on the fingerprint id.

**Comment by harveyyoyo** [comment]
Turned out to be a bad sensor. The bad one was labeled dy50, had only red wires and the light was green. The new one has colored wires and blue light. dont know if it was an incompatible sensor or a faulty one.,

**Pull request merged**